icu4c/source/test/intltest/headeronlytest.cpp

```diff
@@ -0,0 +1,213 @@
+    IcuTestErrorCode errorCode(*this, "TestUSetRangeIterator");
+    using U_HEADER_NESTED_NAMESPACE::USetRanges;
+    using U_HEADER_NESTED_NAMESPACE::CodePointRange;
+    LocalUSetPointer uset(uset_openPattern(u"[abcçカ🚴]", -1, errorCode));
+    std::u16string result;
+    for (auto [start, end] : USetRanges(uset.getAlias())) {
+        // printf("uset.range U+%04lx..U+%04lx\n", (long)start, (long)end);
+        result.append(u" ").append(cpString(start)).append(u"-").append(cpString(end));
+    }
+    assertEquals(WHERE, u" a-c ç-ç カ-カ 🚴-🚴", result);
+    result.clear();
+    for (auto range : USetRanges(uset.getAlias())) {
+        for (UChar32 c : range) {
+            // printf("uset.range.c U+%04lx\n", (long)c);
+            result.append(u" ").append(cpString(c));
+        }
+        result.append(u" |");
+    }
+    assertEquals(WHERE, u" a b c | ç | カ | 🚴 |", result);
+
+    USetRanges range1(uset.getAlias());
+    auto range2(range1);  // copy constructor
+    auto iter = range1.begin();
+    auto limit = range2.end();
+    // operator* with pre- and post-increment
+    {
+        auto cpRange = *iter;
+        assertEquals(WHERE, u'a', cpRange.rangeStart);
+        assertEquals(WHERE, u'c', cpRange.rangeEnd);
+        assertEquals(WHERE, 3, cpRange.size());
+        auto cpRange2(cpRange);
+        auto cpIter = cpRange.begin();
+        auto cpLimit = cpRange2.end();
+        assertEquals(WHERE, u'a', *cpIter++);
+        assertEquals(WHERE, u'b', *cpIter);
+        assertTrue(WHERE, cpIter != cpLimit);
+        CodePointRange::iterator cpIter2(u'b');  // public constructor
+        assertTrue(WHERE, cpIter == cpIter2);
+        assertEquals(WHERE, u'c', *++cpIter);
+        assertTrue(WHERE, cpIter != cpIter2);
+        assertTrue(WHERE, ++cpIter == cpLimit);
+    }
+    ++iter;
+    auto iter2(iter);  // copy constructor
+    assertEquals(WHERE, u'ç', (*iter2).rangeStart);
+    assertEquals(WHERE, u'ç', (*iter2).rangeEnd);
+    assertEquals(WHERE, 1, (*iter2).size());
+    assertEquals(WHERE, u'ç', (*iter2++).rangeStart);
+    assertEquals(WHERE, u'カ', (*iter2).rangeStart);
+    assertTrue(WHERE, ++iter2 != limit);
+    auto iter3(iter2++);
+    assertEquals(WHERE, U'🚴', (*iter3).rangeStart);
+    assertTrue(WHERE, iter2 == limit);
+
+    {
+        CodePointRange cpRange(u'h', u'k');  // public constructor
+        // FYI: currently no operator==
+        assertEquals(WHERE, u'h', cpRange.rangeStart);
+        assertEquals(WHERE, u'k', cpRange.rangeEnd);
+        assertEquals(WHERE, 4, cpRange.size());
+        assertEquals(WHERE, u'i', *++(cpRange.begin()));
+    }
+}
+
+void HeaderOnlyTest::TestUSetStringIterator() {
+    IcuTestErrorCode errorCode(*this, "TestUSetStringIterator");
+    using U_HEADER_NESTED_NAMESPACE::USetStrings;
+    LocalUSetPointer uset(uset_openPattern(u"[abcçカ🚴{}{abc}{de}]", -1, errorCode));
+    std::u16string result;
+    for (auto s : USetStrings(uset.getAlias())) {
+        // UnicodeString us(s);
```

Comment:
This should be `std::u16string` now, right? (Also the same further down.)

Reply:
no: `auto s` is now a std::u16string, so in order to make the .toUTF8String() two lines below work, we need to convert it to a UnicodeString

previously, in usettest.cpp, this line wasn't there, because the s was already a UnicodeString

Reply:
But converting to `UnicodeString` is impossible here, there's a `#define U_SHOW_CPLUSPLUS_API 0` at the beginning of the file, `UnicodeString` doesn't exist here.

Reply:
This is commented-out sample code. Someone who wants to run that, like I did to convince myself that it still worked, has to turn U_SHOW_CPLUSPLUS_API back on. That's what I did for updating the sample code snippets in the API docs.

Reply:
> Someone who wants to run that, like I did to convince myself that it still worked, has to turn U_SHOW_CPLUSPLUS_API back on.

OK … that's something that I did not manage to guess on my own. What do the other reviewers say here: Was this obvious to you?

Reply:
added comments about sample code for pasting